### DISCUSSION TITLE
Espressif ESP32 WROVER E KIT 4.1

### DISCUSSION THREAD

**Discussion by RiggsUS** [HASwitchPlate/openHASP#219] [Hardware]
Hi, I'm looking for help in trying to get openHasp running on a Espressif ESP32 WROVER E KIT 4.1 https://www.adafruit.com/product/3384 I've tried everything I can think off and I just keep hitting a brick wall, I'm a newbe to ESP. 
I've been using the detail from https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-wrover-kit.html for the pinouts for the LCD screen but the most I get is a blank white screen, the back end works ie captive portal.

**Comment by fvanroie** [comment]
Hi there, thanks for checking out openHASP.

Which binary file have you uploaded to the board? The pins for the display are pre-defined in the firmware, so you probably need a custom build that uses the ILI9341 driver and these pins:

ESP32 Pin | LCD Signal
-- | --
GPIO18 | RESET
GPIO19 | SCL
GPIO21 | D/C
GPIO22 | CS
GPIO23 | SDA
GPIO25 | SDO
GPIO5 | Backlight

This translates to a openHASP configuration:

```ini
;region -- TFT_eSPI build options ------------------------
    ${lcd.ili9341}
    -D TFT_MISO=25
    -D TFT_MOSI=23
    -D TFT_SCLK=19
    -D TFT_DC=21
    -D TFT_CS=22
    -D TFT_RST=18       ; RESET
    -D TFT_BCKL=5       ; Default, configurable via web UI (e.g. 21)
;endregion
```

[Customizing the firmware](https://haswitchplate.github.io/openHASP-docs/0.6.1/compiling/gitpod/) and adding a custom [display configuration](https://github.com/HASwitchPlate/openHASP/tree/master/user_setups/esp32) with limited ESP knowledge is a bit like jumping in the deep-end! 😃
For starting out, we generally advise to pick one of the devices or display/dev-board combo's that is supported with a pre-build device firmware.

I'm sure this board can run openHASP, but without having access to the hardware the initial setup is going to be a bit trial-and-error.

One of the reasons we don't have built-in support for this device is because there is no touch capability.
So you can display objects on the screen via MQTT, but you cannot interact with the objects on-screen.

**Reply by RiggsUS** [reply]
I have tried compiling all the combinations in vscode that listed the ILI9341 driver. I thought that board was touchscreen, I must have misread the description.
I have ordered a LOLIN D32 Pro V2.0 and a LOLIN (WEMOS) 320X240 SPI Touch Screen ILI9341 XPT2046 with the cable, from Ali, and have been messing with openHasp on a M5Stack Core 2 AWS.
I thought it would be nice to mess around with the bigger screen on the dev kit while I wait on Ali shipping.

I guess if I do get it running I could add some buttons to do some basic interactions maybe.

**Reply by fvanroie** [reply]
> I have tried compiling all the combinations in vscode that listed the ILI9341 driver.

The pins used by the ESP32 WROVER E KIT are a bit peculiar, I don't think any of the other ILI9341 configurations match those pins.
You will need to change an existing config in the user_setup folder or create a new one. Then compile the modified configuration.
That should work, without touch. Indeed, you can add buttons as inputs too...

**Reply by RiggsUS** [reply]
So if I really just want to time waist and learn I could make a copy of arduitouch-esp32_ili9341 rename it and make the edits for the pins you mentioned in your first comment.

**Reply by fvanroie** [reply]
Yes, that's correct. However I wouldn't rename the file just yet and first edit the existing config, while backing up the original file.
Once it's working you can rename the file *and* change the `[env:arduitouch-esp32_ili9341]` section of the ini file too.

If you make a new file, be sure to also change the `[env:arduitouch-esp32_ili9341]` to `[env:esp32-wrover-e-kit]` or similar.
Then refresh the environments in VScode or close and reopen the application.

Please do share the working config once you get openHASP running.

**Reply by RiggsUS** [reply]
I'm running into this error: openHASP\platformio.ini' (project configuration file): 'No option 'ili9341' in section: 'lcd'' I'm guessing I need to edit this section in the platformio_override.ini 
region -- TFT_eSPI build options -----------------------------------
    ${lcd.lolin24}
    ${esp32.vspi}        ; Use VSPI hardware SPI bus
    -D TFT_DC=5
    -D TFT_CS=26 
    -D TFT_RST=-1        ; RST
    -D TFT_BCKL=22       ; None, configurable via web UI (e.g. 21)
    -D TOUCH_CS=17       ; (can also be 22 or 16)
; -- Options ----------------------------------------

**Reply by RiggsUS** [reply]
I got past the above error by adding the below section to lcd_config.ini

ili9341 =
    -D ILI9341_DRIVER=1
    -D TFT_WIDTH=240
    -D TFT_HEIGHT=320
    -D TFT_ROTATION=0 ; Use default, see TFT_ROTATION values
    -D SPI_FREQUENCY=27000000
    -D SPI_READ_FREQUENCY=20000000
    -D USER_SETUP_LOADED=1
    ;-D TOUCH_DRIVER=610 ;STMPE610
    ;-D SUPPORT_TRANSACTIONS   ; Default on ESP32

But this then leads to a new error:
Compiling .pio\build\arduitouch-esp32_ili9341\src\main_arduino.cpp.o
Compiling .pio\build\arduitouch-esp32_ili9341\src\main_sdl2.cpp.o
src/hasp_gui.cpp: In function 'void guiSetup()':
src/hasp_gui.cpp:302:25: error: 'touch_read' was not declared in this scope
     indev_drv.read_cb = touch_read;
                         ^
Compiling .pio\build\arduitouch-esp32_ili9341\src\mouse_cursor_icon.c.o
*** [.pio\build\arduitouch-esp32_ili9341\src\hasp_gui.cpp.o] Error 1

**Reply by fvanroie** [reply]
Indeed, it's because there is no touch driver defined in this configuration.
I have added a dummy function that fixes this issue: 425886b
Please do a `git pull` to use the new code.

Alternately you can define a touch driver, for example:
```ini
    -D TOUCH_DRIVER=2046     ; XPT2046 Resistive SPI touch panel driver
    -D TOUCH_CS=-1
```

**Comment by fvanroie** [comment]
You can define a new config in `platform_override.ini` too. In that case replace the `[my_custom_config]` section with this:

```ini
; -- Custom Environment configuration example -----------------
[env:my_custom_build]
extends = esp32
board = esp32dev
board_build.partitions = user_setups/esp32/partitions_4MB.csv ; default.csv

upload_port = COM3
monitor_port = COM3
monitor_filters = esp32_exception_decoder

debug_tool = esp-prog
debug_init_break = tbreak setup

build_flags =
    ${env.build_flags}
    ${esp32.build_flags}
;region -- TFT_eSPI build options ------------------------
    ${lcd.ili9341}
    -D TFT_MISO=25
    -D TFT_MOSI=23
    -D TFT_SCLK=19
    -D TFT_DC=21
    -D TFT_CS=22
    -D TFT_RST=18       ; RESET
    -D TFT_BCKL=5       ; Default, configurable via web UI (e.g. 21)
;endregion

lib_deps =
    ${env.lib_deps}
    ${esp32.lib_deps}

lib_ignore =
    ${env.lib_ignore}
    ${esp32.lib_ignore}
```

Make sure you have a file `user_setups\lcd_config.ini`. The line `${lcd.ili9341}` is just short for the following defines in that file:

```ini
    -D ILI9341_DRIVER=1
    -D TFT_WIDTH=240
    -D TFT_HEIGHT=320
    -D TFT_ROTATION=0 ; Use default, see TFT_ROTATION values
    -D SPI_FREQUENCY=40000000
    -D SPI_READ_FREQUENCY=20000000
    -D USER_SETUP_LOADED=1
    -D SUPPORT_TRANSACTIONS
```

If it can't find ${lcd.ili9341}, then just replace the line with those extra defines in your `platform_override.ini`.
The environment to build is called `my_custom_build` in this case.

**Reply by RiggsUS** [reply]
I saw your comment right after I posted mine and followered your directions but I still get this error:
Compiling .pio\build\my_custom_build\src\main_arduino.cpp.o
src/hasp_gui.cpp: In function 'void guiSetup()':
src/hasp_gui.cpp:302:25: error: 'touch_read' was not declared in this scope
     indev_drv.read_cb = touch_read;
                         ^
Compiling .pio\build\my_custom_build\src\main_sdl2.cpp.o
*** [.pio\build\my_custom_build\src\hasp_gui.cpp.o] Error 1

**Reply by fvanroie** [reply]
It means you're not at the latest version of the code. You need to pull the latest version again with the update ` src/drv/touch/touch_driver.h` file.

Do a `git pull` or clone the project again.

**Reply by RiggsUS** [reply]
I had only just pulled it after my first comment with git clone --recursive https://github.com/HASwitchPlate/openHASP then  git checkout 0.6.1 also did git submodule update --init --recursive.
just did git pull again and it says "Already up to date."

**Reply by fvanroie** [reply]
You need to change to the current development branch with `git checkout master`.
0.6.1 is fixed after release. 0.6.2 is being prepared in the `master` branch but not released yet.

**Reply by RiggsUS** [reply]
Sorry for the silence, so far the unite boots with no backlight until it receives  "MQTT PUB: backlight => {"state":"on","brightness":0}" then the backlight turns on but pages display I'm trying to fiddle around with the pages right now to see if its something I have done to stop pages displaying

**Reply by fvanroie** [reply]
Good to hear that you are making progress!

> no backlight until it receives "MQTT PUB: backlight => {"state":"on","brightness":0}"

That sounds to me that the backlight pin logic is reversed. Try adding this line to your configuration:

```ini
    -D TFT_BACKLIGHT_ON=LOW
```

**Reply by RiggsUS** [reply]
I just got pages to show, I will now try adding the above line and recompile.

**Reply by RiggsUS** [reply]
Question did you include somewhere in the firmware the ability to add a rotary encoder to the gpio.

**Reply by fvanroie** [reply]
There is no rotary encoder support currently in the firmware.

**Comment by RiggsUS** [comment]
Ok I'll have to have a play and see if I can figure out how its done.
And here is the config I have so far:
```
;***************************************************;
;           ESP32 Wrover E Devkit with 2.4"         ;
;               - ESP32 WROVER board                ;
;               - ili9341 TFT                       ;
;                                                   ;
;***************************************************;

[env:esp32_wrover_e_devkit]
extends = esp32
board = esp32dev
board_build.partitions = user_setups/esp32/partitions_4MB.csv ; default.csv

upload_port = COM9
monitor_port = COM9
monitor_filters = esp32_exception_decoder

debug_tool = esp-prog
debug_init_break = tbreak setup

build_flags =
    ${env.build_flags}
    ${esp32.build_flags}
;region -- TFT_eSPI build options ------------------------
    ${lcd.ili9341}
    -D TFT_MISO=25
    -D TFT_MOSI=23
    -D TFT_SCLK=19
    -D TFT_DC=21
    -D TFT_CS=22
    -D TFT_RST=18       ; RESET
    -D TFT_BCKL=5       ; Default, configurable via web UI (e.g. 21)
    -D TFT_BACKLIGHT_ON=LOW
;endregion

;region -- Library options -------------------------------
lib_deps =
    ${env.lib_deps}
    ${esp32.lib_deps}

lib_ignore =
    ${env.lib_ignore}
    ${esp32.lib_ignore}
;endregion
```
